### PR TITLE
Entity-view update with the same schema using a csv file path 

### PR DESCRIPTION
**Pull request by teslajoy**

### CODE REVIEW

File: synAnnotationUtils/update.py [hunk 1]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+    df = pandas.read_csv(path, index_col=indexCol).dropna(how='all')
+    df = df.dropna(axis=1, how='all')
+
+    return df
+
+
+def _query2df(syn, synId, clause):
+    """
+    Downloads the entity-view query result as a CsvFileTable then convert it into a pandas dataframe
+    with ROW_IDs as indecies
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: indicies (sp)

File: synAnnotationUtils/update.py [hunk 2]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+    """
+    df = pandas.read_csv(path, index_col=indexCol).dropna(how='all')
+    df = df.dropna(axis=1, how='all')
+
+    return df
+
+
+def _query2df(syn, synId, clause):
+    """
+    Downloads the entity-view query result as a CsvFileTable then convert it into a pandas dataframe
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: file view (not entity view)

File: synAnnotationUtils/update.py [hunk 3]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+
+    :param syn:         A Synapse object: syn = synapseclient.login(username, password) - Must be logged into synapse
+    :param synId:       A Synapse ID of an entity-view (Note: Edit permission on its' files is required)
+    :param clause:      A SQL clause to allow for sub-setting & row filtering in order to reduce the data-size on dow
+    :return:            A dataframe with the entity-view schema and content
+    """
+    query = 'select * from '
+
+    if clause is None:
+        view = syn.tableQuery(query + synId)
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: could also use join, but this is clear enough

File: synAnnotationUtils/update.py [hunk 4]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+    else:
+        view_df.update(user_df)
+
+    schema_unmatch = list(user_df.columns[~user_df.columns.isin(view_df.columns)])
+
+    if not schema_unmatch:
+        logging.info("Updated annotations on entity-view %s." % synId)
+        view = syn.store(synapseclient.Table(synId, view_df))
+    else:
+        logging.info("Updated annotation's dataframe and entity-view's %s schemas don't match." % synId)
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: output the column names that don't match

File: synAnnotationUtils/update.py [hunk 5]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+
+    user_df = _csv2df(path, indexCol)
+
+    minimal_view_schema_column_names = [x['name'] for x in syn.restGET("/column/tableview/defaults/file")['list']]
+
+    # Find existing and new fields/columns in users entity-views schema
+    new_cols = list(user_df.columns[~user_df.columns.isin(minimal_view_schema_column_names)].dropna())
+    col_types = pandas.DataFrame(user_df[new_cols].dtypes)
+    col_types = col_types.replace(['object', 'int64', 'float64'], ['STRING', 'INTEGER', 'FLOAT'])
+
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: Retain synapse type for columns that are pre-existing in the schema.

File: synAnnotationUtils/update.py [hunk 6]
```diff
@@ -64,154 +65,301 @@ def updateAnnoByIdDictFromDict(syn,idDict,annoDict,forceVersion = False):
...
+    if clause is None:
+        new_view = syn.tableQuery('select * from ' + schema.id)
+    else:
+        new_view = syn.tableQuery('select * from ' + schema.id + ' ' + clause)
+
+    new_view = new_view.asDataFrame()
+
+    schema_unmatch = list(user_df.columns[~user_df.columns.isin(new_view.columns)])
+
+    if not schema_unmatch:
```
Comment (kkdang) [inline synAnnotationUtils/update.py]: Can this code be helper function?

File: synAnnotationUtils/update.py (hunk 1 above)
Comment (teslajoy) [inline synAnnotationUtils/update.py]: done!

File: synAnnotationUtils/update.py (hunk 2 above)
Comment (teslajoy) [inline synAnnotationUtils/update.py]: decided to go with entityview which has the same naming convention as the python client schema class type.

File: synAnnotationUtils/update.py (hunk 3 above)
Comment (teslajoy) [inline synAnnotationUtils/update.py]: changing to join

File: synAnnotationUtils/update.py (hunk 4 above)
Comment (teslajoy) [inline synAnnotationUtils/update.py]: done!

**Review by kdaily — APPROVED** [review]
Looks good, but undo the code to handle indices later.

**Pull request merged**